python/lsst/ip/isr/linearize.py

```diff
@@ -26,226 +26,510 @@
 from lsst.pipe.base import Struct
 from .applyLookupTable import applyLookupTable
 
-__all__ = ["LinearizeBase", "LinearizeLookupTable", "LinearizeSquared"]
+__all__ = ["Linearizer",
+           "LinearizeBase", "LinearizeLookupTable", "LinearizeSquared",
+           "LinearizeProportional", "LinearizePolynomial", "LinearizeNone"]
 
 
-def getLinearityTypeByName(linearityTypeName):
-    """Determine the linearity class to use if the type is known.
+class Linearizer(abc.ABC):
+    """Parameter set for linearization.
+
+    These parameters are included in cameraGeom.Amplifier, but
+    should be accessible externally to allow for testing.
 
     Parameters
     ----------
-    linearityTypeName : str
-        String name of the linearity type that is needed.
-
-    Returns
-    -------
-    linearityType : `~lsst.ip.isr.linearize.LinearizeSquared`
-        The appropriate linearity class to use.  If no matching class
-        is found, `None` is returned.
+    table : `numpy.array`, optional
+        Optional lookup table to use for correction.
+    override : `bool`, optional
+        Override the parameters defined in the detector/amplifier.
+    log : `lsst.log.Log`, optional
+        Logger to handle messages.
     """
-    for t in [LinearizeLookupTable, LinearizeSquared]:
-        if t.LinearityType == linearityTypeName:
-            return t
-    return None
+    def __init__(self, table=None, detector=None, override=False, log=None):
+        self._detectorName = None
+        self._detectorSerial = None
+
+        self.linearityCoeffs = dict()
+        self.linearityType = dict()
+        self.linearityThreshold = dict()
```

Comment:
The Threshold and Maximum are values that are defined in the amplifiers in afw.cameraGeom.  The original idea was to have the Threshold be a minimum value below which linearity wasn't needed, and the Maximum to be the maximum value above which the correction was unreliable.  However, no Threshold value has ever been set to anything but 0.0, and the SUSPECT flag is now the default way to mark pixels that are high but not saturated.

The BBox is unfortunately needed to avoid passing an amplifier object around.  The box is still loaded generally loaded from the amplifier.